pyro/poutine/broadcast_messenger.py

```diff
@@ -38,47 +41,51 @@ class BroadcastMessenger(Messenger):
 
     @staticmethod
     @ignore_jit_warnings(["Converting a tensor to a Python boolean"])
-    def _pyro_sample(msg):
+    def _pyro_sample(msg: Message) -> None:
         """
         :param msg: current message at a trace site.
         """
-        if msg["done"] or msg["type"] != "sample":
+        if (
+            msg["done"]
+            or msg["type"] != "sample"
+            or not isinstance(msg["fn"], TorchDistributionMixin)
```

Comment:
Why not literally `getattr(dist, "batch_shape", None) is None` or `hasattr(dist, "batch_shape")`?  I'm wary of changing behavior that is exercised by every Pyro model of every user.  

I do like the return-early refactoring 👍

Reply:
The code also expects `dist` to have `has_rsample` and `expand` attributes.  `has_rsample` is `pyro.distributions.Distribution`'s attribute and `expand` is `TorchDistributionMixin`'s attribute. I thought having these three attributes (`batch_shape`, `has_rsample`, `expand`) qualifies the `dist` to be a `TorchDistributionMixin` instance. I know this makes the code stricter and there is some chance (although I think very small) that it might break someone's code. I'm okay changing it back to `getattr(dist, "batch_shape", None) is None` and also checking for `has_rsample` and `expand` attributes if you think it is better to keep it that way. I leave it up to you to decide 😄 .

Reply:
Thanks for explaining. Probably there are very few or no places where `msg["fn"]` is not a `TorchDistributionMixin`, and I think the benefits of type checking outweigh that possible breakage. So as-is is fine 👍